plugins/Validation/src/Validation.ts

```diff
@@ -1,223 +1,346 @@
+/* tslint:disable:no-any */
+import { State } from '@hookstate/core';
+import { PluginCallbacks, StateValueAtRoot } from '@hookstate/core/dist';
 
-import { Plugin, Path, StateValueAtPath, State } from '@hookstate/core';
+export const ValidationId = Symbol('Validation');
 
-export type ValidationSeverity = 'error' | 'warning';
+type ValidateFn<T> = (value: T) => boolean;
+type Path = readonly (string | number | symbol)[];
 
-export interface ValidationError {
-    readonly message: string;
-    readonly path: Path;
-    readonly severity: ValidationSeverity;
+interface NestedState {
+    state: State<any>;
+    parent?: NestedState;
 }
 
-export interface ValidationExtensions<S> {
-    validate(attachRule: (value: S) => boolean,
-        message: string | ((value: S) => string),
-        severity?: ValidationSeverity): void,
-    validShallow(): boolean,
-    valid(): boolean,
-    invalidShallow(): boolean,
-    invalid(): boolean,
-    firstError(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number
-    ): Partial<ValidationError>,
-    errors(
-        filter?: (e: ValidationError) => boolean,
-        depth?: number,
-        first?: boolean
-    ): ReadonlyArray<ValidationError>,
+interface SingleValidator<T> {
+    required(message?: string): void;
+
+    when(fn: (value: State<T>) => boolean): this;
+
+    validate(validator: ValidateFn<T>, message?: string): void;
 }
 
-const PluginID = Symbol('Validate');
+type NestedValidator<Root, T> = T extends string ? SingleValidator<T> :
```

Comment:
Why only `string`? where are numbers, booleans, undefined and null?

Reply:
why undefined/null?